upgrade-1.x-to-2.x.sh

```diff
@@ -0,0 +1,410 @@
+        printf "[%09d%s%s\n" "$(($ENDTIME - $STARTTIME))" "][$loglevel]" "$1" | tee -a $LOGFILE
+    else
+        printf "[%09d%s%s\n" "$(($ENDTIME - $STARTTIME))" "][$loglevel]" "$1"
+    fi
+    if [ "$loglevel" == "ERROR" ]; then
+        /usr/bin/resin-device-progress --percentage 100 --state "ResinOS: Update failed."
+        exit 1
+    fi
+}
+
+# Test if a version is greater than another
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Log timer
+STARTTIME=$(date +%s)
+
+# Check board support
+case $SLUG in
+    beaglebone*)
+	TARGET_VERSION=2.0.7_rev2
+	# In 2.x there is only a single device type
+	DEVICE=beaglebone-black
+	BINARY_TYPE=arm
+	;;
+    raspberry*)
+	TARGET_VERSION=2.0.6_rev3
+	DEVICE=$SLUG
+	BINARY_TYPE=arm
+	;;
+    *)
+	log ERROR "Unsupported board type $SLUG."
+esac
+
+# Check host OS version
+if version_gt $VERSION $MIN_HOSTOS_VERSION || [ "$VERSION" == $MIN_HOSTOS_VERSION ]; then
+    log "Host OS version $VERSION OK."
+else
+    case $VERSION in
+	1.*)
+	    log ERROR "Host OS version \"$VERSION\" too low, need \"$MIN_HOSTOS_VERSION\" or greater."
+	    ;;
+	*)
+	    log ERROR "Host OS version \"$VERSION\" too high."
+	    ;;
+    esac
+fi
+
+# Check we are using the first root filesystem
+root_part=$(findmnt -n --raw --evaluate --output=source /)
+case $root_part in
+    *p2)
+	log "Current root partition is first root partition."
+	;;
+    *p3)
+	log "Current root partition $root_part is the second root partition. Copying existing OS to first partition..."
+	root_dev=${root_part%p3}
+	dd if=${root_dev}p3 of=${root_dev}p2
+	log "Updating bootloader to point to first partition..."
+	case $SLUG in
+	    beaglebone*)
+		sed -i -e 's/bootpart=1:3/bootpart=1:2/' /mnt/boot/uEnv.txt
+		;;
+	    raspberry*)
+		sed -i -e 's/mmcblk0p3/mmcblk0p2/' /mnt/boot/cmdline.txt
+		;;
+	esac
+	log "Rebooting..."
+	sync
+	reboot
+	;;
+    *)
+	log ERROR "Current root partition $root_part is not first or second root partition, aborting."
+	;;
+esac
+root_dev=${root_part%p2}
+
+# Check if we need to install some more extra tools
+if ! version_gt $VERSION $PREFERRED_HOSTOS_VERSION && ! [ "$VERSION" == $PREFERRED_HOSTOS_VERSION ]; then
+    log "Host OS version $VERSION is less than $PREFERRED_HOSTOS_VERSION, installing tools..."
+    tools_path=/home/root/upgrade_tools
+    mkdir -p $tools_path
+    export PATH=$tools_path:$PATH
+    case $BINARY_TYPE in
+	arm)
+	    download_uri=https://github.com/resin-os/resinhup/raw/onetotwo/upgrade-binaries/$BINARY_TYPE
+	    curl -s -L -o $tools_path/tar $download_uri/tar
+	    chmod 755 $tools_path/tar
+	    curl -s -L -o $tools_path/e2label $download_uri/e2label
+	    chmod 755 $tools_path/e2label
+	    curl -s -L -o $tools_path/mkfs.ext4 $download_uri/mkfs.ext4
+	    chmod 755 $tools_path/mkfs.ext4
+	    curl -s -L -o $tools_path/resize2fs $download_uri/resize2fs
+	    chmod 755 $tools_path/resize2fs
```

Comment:
Would it be better if this was an array of the tool names, and a loop over those array items with curl + chmod? It would be less duplication and more readable, imho.